File: backend/app/schemas/schedule.py
from __future__ import annotations

from datetime import date as _date, datetime as _datetime
from typing import Optional

from pydantic import BaseModel, Field


# ==================== SchSchedule Schema ====================

class ScheduleCreate(BaseModel):
    """创建排班记录（单日单班次）"""
    date: _date = Field(..., description="排班日期")
    shift_id: int = Field(..., gt=0, description="班次模板ID")
    org_id: int = Field(..., gt=0, description="组织ID")
    leader_staff_id: Optional[int] = Field(None, description="值班领导人员ID")
    status: int = Field(default=0, description="0=草稿 1=已发布 2=已撤回")
    source: str = Field(default="manual", description="生成方式 auto/manual/swap")


class ScheduleUpdate(BaseModel):
    """更新排班记录（仅草稿可更新）"""
    date: Optional[_date] = None
    shift_id: Optional[int] = Field(None, gt=0)
    org_id: Optional[int] = Field(None, gt=0)
    leader_staff_id: Optional[int] = None


# ==================== SchScheduleDetail Schema ====================

class ScheduleDetailCreate(BaseModel):
    """创建排班明细"""
    staff_id: int = Field(..., gt=0, description="人员ID")
    role_type: str = Field(default="member", description="角色类型 leader/member")
    is_substitute: bool = Field(default=False, description="是否替班")
    note: Optional[str] = Field(None, max_length=200, description="备注")


class ScheduleDetailUpdate(BaseModel):
    """更新排班明细"""
    staff_id: Optional[int] = Field(None, gt=0)
    role_type: Optional[str] = None
    is_substitute: Optional[bool] = None
    note: Optional[str] = Field(None, max_length=200)


class AssignStaffRequest(BaseModel):
    """为排班分配人员"""
    staff_id: int = Field(..., gt=0, description="人员ID")
    role_type: str = Field(default="member", description="角色类型 leader/member")
    is_substitute: bool = Field(default=False, description="是否替班")
    note: Optional[str] = None


class RemoveStaffRequest(BaseModel):
    """移除排班中的人员"""
    staff_id: int = Field(..., gt=0, description="人员ID")


class BatchDetailItem(BaseModel):
    """批量操作单条明细"""
    id: Optional[int] = Field(None, description="明细ID，有则更新无则创建")
    schedule_id: int = Field(..., gt=0)
    staff_id: int = Field(..., gt=0)
    role_type: str = "member"
    is_substitute: bool = False
    note: Optional[str] = None


class BatchDetailRequest(BaseModel):
    """批量创建/更新排班明细"""
    items: list[BatchDetailItem]


class BatchPublishRequest(BaseModel):
    """批量发布/撤回排班"""
    schedule_ids: list[int] = Field(..., min_length=1, description="排班记录ID列表")


# ==================== 响应 Schema ====================

class StaffInfo(BaseModel):
    """人员信息"""
    staff_id: int
    name: str
    role_type: str = "member"
    is_substitute: bool = False
    note: Optional[str] = None


class ShiftInfo(BaseModel):
    """班次信息"""
    shift_id: int
    shift_name: str
    shift_color: str = ""
    start_time: str = ""
    end_time: str = ""


class ScheduleDetailResponse(BaseModel):
    """排班明细响应"""
    id: int
    schedule_id: int
    staff_id: int
    staff_name: Optional[str] = None
    role_type: str
    is_substitute: bool = False
    note: Optional[str] = None

    class Config:
        from_attributes = True


class ScheduleResponse(BaseModel):
    """排班记录响应（含明细）"""
    id: int
    date: _date
    shift_id: int
    shift_name: Optional[str] = None
    shift_color: Optional[str] = None
    shift_start_time: Optional[str] = None
    shift_end_time: Optional[str] = None
    org_id: int
    org_name: Optional[str] = None
    leader_staff_id: Optional[int] = None
    leader_name: Optional[str] = None
    status: int
    source: str
    published_at: Optional[_datetime] = None
    published_by: Optional[int] = None
    created_at: Optional[_datetime] = None
    updated_at: Optional[_datetime] = None
    details: list[ScheduleDetailResponse] = []

    class Config:
        from_attributes = True


class ScheduleListResponse(BaseModel):
    """排班列表响应"""
    items: list[ScheduleResponse]
    total: int


# ==================== 日历视图 ====================

class CalendarShift(BaseModel):
    """日历中的单个班次"""
    schedule_id: int
    shift_template_id: int
    shift_name: str
    shift_color: str = ""
    start_time: str = ""
    end_time: str = ""
    leader: Optional[StaffInfo] = None
    members: list[StaffInfo] = []
    status: int = 0
    source: str = "manual"
    conflicts: list[str] = []


class CalendarDate(BaseModel):
    """日历中单天的数据"""
    date: str
    shifts: list[CalendarShift] = []


class CalendarResponse(BaseModel):
    """排班日历响应"""
    dates: list[CalendarDate] = []


# ==================== 人员排班统计 ====================

class StaffShiftStat(BaseModel):
    """人员单日排班信息"""
    date: _date
    shift_name: str
    shift_color: str = ""
    start_time: str = ""
    end_time: str = ""
    role_type: str = "member"


class StaffSummaryResponse(BaseModel):
    """人员排班统计响应"""
    staff_id: int
    total_days: int = 0
    total_hours: float = 0.0
    night_shifts: int = 0
    recent_shifts: list[StaffShiftStat] = []

File: frontend/src/api/schedule.ts
import api from './index'

// ==================== 类型定义 ====================

export interface Schedule {
  id: number
  date: string
  shift_id: number
  shift_name: string | null
  shift_color: string | null
  shift_start_time: string | null
  shift_end_time: string | null
  org_id: number
  org_name: string | null
  leader_staff_id: number | null
  leader_name: string | null
  status: number  // 0=草稿 1=已发布 2=已撤回
  source: string  // auto/manual/swap
  published_at: string | null
  published_by: number | null
  created_at: string | null
  updated_at: string | null
  details: ScheduleDetail[]
}

export interface ScheduleDetail {
  id: number
  schedule_id: number
  staff_id: number
  staff_name: string | null
  role_type: string  // leader/member
  is_substitute: boolean
  note: string | null
}

export interface StaffInfo {
  staff_id: number
  name: string
  role_type: string
  is_substitute?: boolean
  note?: string | null
}

export interface CalendarShift {
  schedule_id: number
  shift_template_id: number
  shift_name: string
  shift_color: string
  start_time: string
  end_time: string
  leader: StaffInfo | null
  members: StaffInfo[]
  status: number
  source: string
  conflicts: string[]
}

export interface CalendarDate {
  date: string
  shifts: CalendarShift[]
}

export interface CalendarResponse {
  dates: CalendarDate[]
}

export interface ScheduleListResponse {
  items: Schedule[]
  total: number
}

export interface ScheduleCreateForm {
  date: string
  shift_id: number
  org_id: number
  leader_staff_id?: number | null
  source?: string
}

export interface ScheduleUpdateForm {
  date?: string
  shift_id?: number
  org_id?: number
  leader_staff_id?: number | null
}

export interface AssignStaffForm {
  staff_id: number
  role_type?: string
  is_substitute?: boolean
  note?: string
}

export interface BatchPublishForm {
  schedule_ids: number[]
}

// ==================== API 函数 ====================

export function getSchedules(params?: {
  org_id?: number
  start_date?: string
  end_date?: string
  status?: number
  shift_id?: number
  page?: number
  page_size?: number
}): Promise<ScheduleListResponse> {
  return api.get('/schedules', { params })
}

export function getScheduleCalendar(params: {
  start_date: string
  end_date: string
  org_id?: number
  status?: number
}): Promise<CalendarResponse> {
  return api.get('/schedules/calendar', { params })
}

export function getSchedule(id: number): Promise<Schedule> {
  return api.get(`/schedules/${id}`)
}

export function createSchedule(data: ScheduleCreateForm): Promise<Schedule> {
  return api.post('/schedules', data)
}

export function updateSchedule(id: number, data: ScheduleUpdateForm): Promise<Schedule> {
  return api.put(`/schedules/${id}`, data)
}

export function deleteSchedule(id: number): Promise<any> {
  return api.delete(`/schedules/${id}`)
}

export function assignStaff(scheduleId: number, data: AssignStaffForm): Promise<any> {
  return api.post(`/schedules/${scheduleId}/assign-staff`, data)
}

export function removeStaff(scheduleId: number, staffId: number): Promise<any> {
  return api.post(`/schedules/${scheduleId}/remove-staff`, { staff_id: staffId })
}

export function publishSchedules(scheduleIds: number[]): Promise<any> {
  return api.post('/schedules/publish', { schedule_ids: scheduleIds })
}

export function recallSchedules(scheduleIds: number[]): Promise<any> {
  return api.post('/schedules/recall', { schedule_ids: scheduleIds })
}

export function recallSchedulesByMonth(orgId: number, year: number, month: number): Promise<any> {
  return api.post('/schedules/recall-month', null, { params: { org_id: orgId, year, month } })
}

export function approveSchedules(scheduleIds: number[]): Promise<any> {
  return api.post('/schedules/approve', { schedule_ids: scheduleIds })
}

export function rejectSchedules(scheduleIds: number[]): Promise<any> {
  return api.post('/schedules/reject', { schedule_ids: scheduleIds })
}

export function getStaffSummary(staffId: number, days?: number): Promise<any> {
  return api.get(`/schedules/staff-summary/${staffId}`, { params: { days } })
}

// ==================== 工作量统计 ====================

export interface ScheduleStatisticsItem {
  staff_id: number
  staff_name: string
  employee_no: string
  org_name: string
  total_shifts: number
  total_hours: number
  night_shifts: number
  weekend_shifts: number
  leader_shifts: number
  holiday_shifts: number
  weight_score: number
}

export interface ScheduleStatisticsSummary {
  total_staff: number
  total_shifts: number
  avg_shifts_per_person: number
  avg_hours_per_person: number
  total_night_shifts: number
  total_holiday_shifts: number
}

export interface ScheduleStatisticsResponse {
  period: { start: string; end: string }
  items: ScheduleStatisticsItem[]
  summary: ScheduleStatisticsSummary
}

export function getScheduleStatistics(params: {
  start_date: string
  end_date: string
  org_id?: number
  top?: number
}): Promise<ScheduleStatisticsResponse> {
  return api.get('/schedules/statistics', { params })
}
